### DISCUSSION TITLE
Speakers not found at inference time

### DISCUSSION THREAD

**Discussion by Ca-ressemble-a-du-fake** [coqui-ai/TTS#2055] [General Q&A]
Hi,

I have 4 datasets each of which spoken by a different speaker. I followed the tutorial about Multi Speaker to train a VITS model. 

So I enclosed the following in my training_4_VOices_py script :

```
# init speaker manager for multi-speaker training
# it maps speaker-id to speaker-name in the model and data-loader
speaker_manager = SpeakerManager()
speaker_manager.set_ids_from_data(train_samples + eval_samples, parse_key="speaker_name")
config.num_speakers = speaker_manager.num_speakers
print( str(speaker_manager.num_speakers) + " speakers detected") => This prints "4 speakers detected"

# init model
model = Vits(config, ap, tokenizer, speaker_manager=speaker_manager)
```

So the speaker manager seems to be well set up but then the test audios (in Tensorboard) sound like a mix of those different voices in each audio (even at step 30k) although different speaker names were specified and at inference time :

```
python3 ~/coquiTTS/TTS/TTS/bin/synthesize.py --list_speaker_idxs --model_path ~/Models/VITS_4_Voices-October-06-2022_03+43PM-0000000/best_model_3861.pth --config_path ~/Models/VITS_4_Voices-October-06-2022_03+43PM-0000000/config.json
```
Returns 

```
 > Available speaker ids: (Set --speaker_idx flag to one of these values to use the multi-speaker model.
Traceback (most recent call last):
  File "/home/caraduf/coquiTTS/TTS/TTS/bin/synthesize.py", line 367, in <module>
    main()
  File "/home/caraduf/coquiTTS/TTS/TTS/bin/synthesize.py", line 326, in main
    print(synthesizer.tts_model.speaker_manager.ids)
AttributeError: 'NoneType' object has no attribute 'ids'
```
Also the speakers.pth shows the speaker names (although this is not a proper text file I could "read" some parts) so it looks like it is working.

However the generated config.json has `num_speakers` field set to 0 whereas the recipe sets it differently (see `config.num_speakers = speaker_manager.num_speakers`).

How can I make those 4 speakers be available for inference ? Am I supposed to provide the speaker embeddings beforehand (although it is not mentioned in the tutorial (for GlowTTS), it may be necessary for VITS) ?

Many thanks in advance :smile:

**Comment by ZimmanJ** [comment]
hi, i am having the exact same problem .. Did you figure it out in the end ?

**Reply by Ca-ressemble-a-du-fake** [reply]
Hi, I can't remember and stopped doing this. Did you try to update to the latest coqui version. I can't remember for sure but they may have fixed this (or I mix this with an other question)!

**Reply by ZimmanJ** [reply]
Ok.. thank u

**Reply by Ca-ressemble-a-du-fake** [reply]
Oh! I googled the problem just now because I am back to multi speaker and found your other [question](https://github.com/coqui-ai/TTS/discussions/2330) :smile: !

**Reply by Ca-ressemble-a-du-fake** [reply]
@ZimmanJ Now I've figured it out. It's [here](https://github.com/coqui-ai/TTS/discussions/2217#discussioncomment-4936201). So basically you've got to compute the embeddings for all your speakers with `bin/compute_embeddings.py` and then put the path to this speakers.pth file in the config.json associated with your model (`d_vector_file = "path_to_speakers.pth"`). Take care not to use a list, just a string.

Please note: it seems like this is needed only for inference. At training a list of pth/json file is needed (as generated by the recipe).

**Reply by ZimmanJ** [reply]
Ohh thank you..  i was just using speaker encoder, instead of using speaker encoding. Greatly appreciated.. 

> @ZimmanJ Now I've figured it out. It's [here](https://github.com/coqui-ai/TTS/discussions/2217#discussioncomment-4936201). So basically you've got to compute the embeddings for all your speakers with `bin/compute_embeddings.py` and then put the path to this speakers.pth file in the config.json associated with your model (`d_vector_file = "path_to_speakers.pth"`). Take care not to use a list, just a string.
> 
> Please note: it seems like this is needed only for inference. At training a list of pth/json file is needed (as generated by the recipe).
